pydantic_ai_slim/pydantic_ai/_parts_manager.py

```diff
@@ -99,58 +179,385 @@ def handle_text_delta(
         Raises:
             UnexpectedModelBehavior: If attempting to apply text content to a part that is not a TextPart.
         """
-        existing_text_part_and_index: tuple[TextPart, int] | None = None
+        potential_part: _ExistingPart[TextPart] | _ExistingPart[ThinkingPart] | None = None
 
         if vendor_part_id is None:
             # If the vendor_part_id is None, check if the latest part is a TextPart to update
             if self._parts:
                 part_index = len(self._parts) - 1
                 latest_part = self._parts[part_index]
                 if isinstance(latest_part, TextPart):
-                    existing_text_part_and_index = latest_part, part_index
+                    potential_part = _ExistingPart(part=latest_part, index=part_index, found_by='latest_part')
+                    # ✅ vendor_part_id and ✅ potential_part is a TextPart
+                else:
+                    # NOTE that the latest part could be a ThinkingPart but
+                    #   -> C4: we require ThinkingParts come from/with vendor_part_id's
+                    # ❌ vendor_part_id is None + ❌ potential_part is None -> new part!
+                    pass
+            else:
+                # ❌ vendor_part_id is None + ❌ potential_part is None -> new part!
+                pass
         else:
             # Otherwise, attempt to look up an existing TextPart by vendor_part_id
             part_index = self._vendor_id_to_part_index.get(vendor_part_id)
             if part_index is not None:
                 existing_part = self._parts[part_index]
```

Comment:
I wonder if a helper method like `_existing_part_by_vendor_id()` could be helpful